### ISSUE TITLE
[Feature Request] Manual log_prob() gradients

### ISSUE THREAD

**Issue by anonymizedrepos** [tensorflow/probability#192]
It seems like using tf.gradients() to differentiate through a log_prob can be significantly slower than computing your own closed-form derivative. I've attached code I wrote for comparison. It uses gradient ascent on the mean to maximize the log probability of a normal.

Is it reasonable to add closed-form gradients with respect to distribution parameters? I'm imagining doing the override in the log_prob() function. If so I'd be happy to start writing them, but I could use some guidance getting started.

```python
import tensorflow as tf
import tensorflow_probability as tfp
tfd = tfp.distributions
import numpy as np
import time
#%%
g = tf.Graph()
sess = tf.InteractiveSession(graph = g)
#%%

np.random.seed(1)
tf.set_random_seed(1)

dim = 100
num_samples = 10000
n_iters = 5000
tf_type_var = tf.float32


true_mean = np.random.normal(loc = np.zeros(dim),scale = np.ones(dim))

np_samples = np.random.normal(loc = true_mean,scale = np.ones(dim),size = [num_samples,dim])
sample_mean = np.mean(np_samples,axis = 0)
np_mean = np.zeros(dim)

def np_gradient(mu,samples): 
    return 2*(np.sum(samples-mu.T,axis =0))


#Test numpy time
t = time.time()
for i in range(0,n_iters):
    np_mean += (1/(num_samples*5))*np_gradient(np_mean,np_samples)

np_time = time.time()-t



tf_samples = tf.constant(np_samples,dtype = tf_type_var)
autodiff_tf_mean = tf.Variable(tf.zeros(dim),dtype = tf_type_var)
manualdiff_tf_mean = tf.Variable(tf.zeros(dim),dtype = tf_type_var)

to_diff = tfd.Normal(loc = autodiff_tf_mean,scale = tf.ones(dim)).log_prob(tf_samples)
grad = tf.gradients(to_diff,autodiff_tf_mean)

autodiff_update_mean = tf.assign_add(autodiff_tf_mean,grad[0]/(num_samples*5))

def tf_gradient(mu,samples):
    return 2*(tf.reduce_sum(samples-mu,axis=0))#,axis = 0))

manualdiff_update_mean = tf.assign_add(manualdiff_tf_mean,tf_gradient(manualdiff_tf_mean,tf_samples)/(num_samples*5)) 

init = tf.global_variables_initializer()
sess.run(init)

t = time.time()
for i in range(0,n_iters):
    sess.run(manualdiff_update_mean)

manualdiff_time = time.time()-t


t = time.time()
for i in range(0,n_iters):
    sess.run(autodiff_update_mean)
autodiff_time = time.time()-t

print('Autodiff tensorflow time ',autodiff_time)
print('Manual diff tensorflow time ',manualdiff_time)
print('Numpy time ',np_time)
#%%
test1 = sess.run(tf.norm(autodiff_tf_mean-manualdiff_tf_mean))<1e-6
test2 = sess.run(tf.norm(autodiff_tf_mean-np_mean))<1e-6
test3 = sess.run(tf.norm(autodiff_tf_mean-sample_mean))<1e-6

if test1 & test2 &test3:
    print("All means agree with sample mean")
else:
    print("Either take more gradient steps or something is wrong")

#print("Autodiff tensorflow mean",sess.run(autodiff_tf_mean))
#print("Manual diff tensorflow mean",sess.run(manualdiff_tf_mean))
#print("Numpy mean",np_mean)
```

**Comment by brianwa84** [comment]
If the gradients are correct (unbiased) and significantly faster, or more
numerically stable, or lower variance, then it's reasonable to override the
default.

For example, see here:
https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/distributions/von_mises.py#L303

**Comment by srvasude** [comment]
How much of a time difference have you observed in practice? I guess running this on my desktop gives me 3x speed up, which is somewhat expected (since gradients are typically 3x expensive).

As Brian mentioned, closed form derivatives can be implemented via a tf.custom_gradient for cases where the overriden gradient is much better behaved then the one autodiff produces. 

The part I am not sure about is whether writing the closed form, and using tf.custom_gradient to override would preserve a speed up (i.e. not sure if you still incur the overhead of autodiff), and if not, whether it makes sense to expose in the Distribution API gradients for each log_prob (for which I would argue against, since it makes the API a lot more complicated).

**Comment by anonymizedrepos** [comment]
I get 4-5x speedup on my CPU as the problem gets larger for the test above. For example 20sec vs 100sec.

For small problems the speed difference isn't noticeable.

However my example ignores any overhead from the gradient override, which I didn't consider.

**Comment by brianwa84** [comment]
The gradient you are computing looks like it assumes the scale is 1. Can you time it with a non-1 scale?
